src/framework/components/rigid-body/system.js

```diff
@@ -552,6 +560,242 @@ class RigidBodyComponentSystem extends ComponentSystem {
         return results;
     }
 
+    /**
+     * Perform a collision check on the world and return all entities the shape hits.
+     * It returns an array of {@link RaycastResult}, one for each hit. If no hits are
+     * detected, the returned array will be of length 0.
+     *
+     * @param {object} shape - The shape to use for collision.
+     * @param {number} shape.axis - The local space axis with which the capsule, cylinder or cone shape's length is aligned. 0 for X, 1 for Y and 2 for Z. Defaults to 1 (Y-axis).
+     * @param {Vec3} shape.halfExtents - The half-extents of the box in the x, y and z axes.
+     * @param {number} shape.height - The total height of the capsule, cylinder or cone from tip to tip.
+     * @param {string} shape.type - The type of shape to use. Available options are "box", "capsule", "cone", "cylinder" or "sphere". Defaults to "box".
+     * @param {number} shape.radius - The radius of the sphere, capsule, cylinder or cone.
+     * @param {Vec3} position - The world space position for the shape to be.
+     * @param {Vec3} rotation - The world space rotation for the shape to have.
+     *
+     * @returns {RaycastResult[]} An array of shapecast hit results (0 length if there were no hits).
+     */
+    shapeCast(shape, position, rotation) {
```

Comment:
So this is a utility function that basically calls the shape-specific function. For the purpose of maintaining a nice and accessible API, should we:

1. just retain this 'all-in-one' function as public
2. just retain the shape-specific function as public
3. keep all public

Reply:
The advantage of "all-in-one" function is that you can actually use a JSON format for all the shape config meaning it can easily be assigned from a script attribute to edit it within the editor directly and avoid having to be only code-based. In another hand the shape-specific function allow users used to other engines to find the function easily (especially `sphereCast` which is really common).

Reply:
Hmm, that's a great point. OK, fair enough. 😄

Reply:
I wanted to unresolve this discussion and talk about it some more, @MushAsterion. I do still feel that this PR introduces a lot of additions to the API. 18 methods at the moment. And if `shapeCastAll` flavors get added, it takes it to 24. Plus the two existing raycasting functions, that'd be 26 functions for testing and casting. I think that's just too much. We _could_ just have:

```
raycast (we could deprecate raycastFirst and raycastAll - note we don't _have_ to do that in this PR)
shapeCast
shapeTest
```

The all versus first could just be an option (maybe defaulting to first).

It just feels to me that the JSON approach is interesting, but it's very niche and, in reality, the vast majority of developers won't call the API like that.

Reply:
You could potentially also drop `shapeTest` if start and end pos/rot of `shapeCast` where the same too maybe?

Reply:
I agree that taking the type of result (ie "all" and "first") could be an option, but wouldn't that actually mess with IntelliSense and other suggestion systems? I mean it could be declared through a .d.ts file but I'm far from an expert in this field and not sure how it would behave at build time...

Honestly the point of the `shapeCast`/`shapeTest` with JSON was just to provide an "all-in-one" solution but that might not be for everyone as you mentioned especially since most of the API is high level, hence the function for each type separately. I tried to follow along with other exposed functions of the API and looked at some other engine to find the best solution and actually most engines only have sphere and sometimes box hence not a big deal to have them separated.

> You could potentially also drop `shapeTest` if start and end pos/rot of `shapeCast` where the same too maybe?

I'm definitely not sure on how Ammo threat such case but I feel like testing whether the values are equal to run another function could impact performance through that extra check at engine-level for a project-level issue? I mean position is always a Vec3 making it easy but rotation accepts Vec3 and Quat making it more conversion before comparing

Reply:
Back on making "all" and "first" an option wouldn't that confuse people that only "first" is supported for `shapeCast`?

Reply:
> I don't see how. The `playcanvas.d.ts` TypeScript declarations would specify the option and Intellisense would check it.
My concern is not about the options type but the returned value type as it returns `HitResult|null` for first values and `HitResult[]` for all. As as my knowledge of JSDoc it's not possible to document as-is and required a .d.ts override to declare a function override for any documentation to pick correctly.

> Another suggestion would be to simply not specify and end pos/rot. This would suggest that it's a test, not a cast.
That would imply setting end position/rotation as an option, do you think it would be intuitive enough? Also it would only work again on the "first" result for shape casting. Do you think documentation would be enough to guide people?

Reply:
Just a few thoughts:

1. Does shapeTestFirst even make sense? My understanding of this is that it is a point in time test for all intersections with the shape. Which result is "first" doesn't seem intuitive, what defines the order? If this is the case, then I'd suggest just dropping First/All methods and always return an array.
2. It looks like shapeCastAll doesn't exist, is this intentional?
3. I think the distinction between raycastFirst and raycastAll is historical because the original version of Ammo didn't support all results. However, it does force the user to make the distinction because (I think) they primarily only want the first result. So I actually don't mind them being separate methods. If they are combined my preference would probably be for a raycast() method would be: a) always return an array of results b) make abort-on-first-hit the default behavior, which does have the downside of making the default return always a length 1 array, but I think provides overall consistency. You don't want people having to check if the return result is null, object or array.

Reply:
1. For `shapeTestFirst` the first result is the closest to the position of the shape. Which is not always the first returned by `shapeTestAll`. Always returning an array would make it a different experience from `raycastFirst`. But if standardizing everything it could actually make sense.

2. `shapeCastAll` requires additional exposure of the Ammo.js library, probably the reason why it was not directly added to the PR. I'm actually not even sure to have managed to make it in any of my builds, but if you manage to make it feel free to contribute or share the knowledge.

3. For sure making both results an array would make everything easier on certain projects but will it be clear for everyone without deep knowledge especially when your option is set to return first result only but the function actually returns an array of length = 1|0 instead of a single object or `null` ?

Reply:
Right, so shapeTestFirst is actually doing more work than shapeTestAll (because it gets all contacts, sorts them and then discards all but the first), which is counter-intuitive if you understand what raycastFirst/All is doing. 

Do you think that use-case is common enough to have an explicit engine method? We could just support shapeTestAll and have the user do the sort/discard if they only need the closest result.

Reply:
And by the way, thank you for this PR! I appreciate the hard work! I found it, because I want to use shape casting in my project.

Reply:
`shapeTestFirst` is doing more work due to how Bullet/Ammo work and completely different from raycasting as raycasting is only a question of point while the closest point in a shape's position is not always the first to be detected since it depends on how the physics engine perform the test.

Once again the exposure of First/All methods is to match with what the engine has already implemented. I'm an external contributor so I don't want to step or change core project features. I personally find it handy for beginners to have the support for sure but as @willeastcott mentioned it's a very long list of new public methods which might also bring more confusion. I think it's a good solution to merge as much as possible to reduce public methods but I'm concerned about how easy it would be for people that are just looking to make a quick project without trying to get a deep understanding of the engine's API... I'll surely update to fit whatever they think is best anyway, the core features of the PR are already done and updating them is not a big deal

Reply:
Good stuff! Many useful features are being added. Although, I do agree that the API is too verbose. We can simplify those, and move most of the customization stuff to options.

My vote goes for dropping -all/-first from the method name, and use an option setting. 

`shapeTestFirst` seems a bit superfluous. We can just return everything and let user decide what to do with the results. If future Ammo adds a feature to allow collecting only the first intersection (which sounds strange to me), we can add it via an option. Shape and point collision tests are usually useful for getting all colliders they intersect, otherwise a sweep test is used.

About shape target rotation during sweep. Personally, I've never met a need to use it. Maybe it is useful in robotics, where Bullet engine is popular? No idea. Most of the other other physics engines don't allow it (Rapier, PhysX, Jolt). As such, I would actually move it to options as well.

Whatever the names are, but signatures could look something like:
```
raycast(start, end, opts = {});

shapecast(shape, pos, rot, dir = Vec3.ZERO, opts = {});
```
If `dir` is a zero vector, we do shape test. Otherwise it is a non-normalized cast direction (its normal for direction, its magnitude as cast distance). Shape properties can be part of the options, e.g. radius for radius-based shapes.

About destroying a shape used in sweep. If a shape is not auto-destroyed (e.g. via some option), then there should be a method to destroy it later.

Reply:
I'm not sure to understand the point of changing `endPosition` into `dir` while `raycast` has a `end` param, is there any particular reason to make it different?

Something like
```JS
shapecast(shape, start, end, opts = {})
```

Where options could have something like

```JS
{
    startRotation: Vec3|Quat,
    endRotation: Vec3|Quat,
    findAll: boolean, // Defaults to false, not possible if end transform !== start transform
    // ... other options
}
```

Would be more fitting with the suggested `raycast` signature.

Reply:
Yep, something like that. I think starting rotation is probably not optional, as it is often used for non-spheres. Target rotation can be optional.

Reply:
Alright so based on feedback I just rewrote the PR. Let me know if it better fits now 🙃